src/os/hostname.zig

```diff
@@ -6,6 +6,91 @@ pub const HostnameParsingError = error{
     NoSpaceLeft,
 };
 
+fn isUriPathSeparator(c: u8) bool {
+    return switch (c) {
+        '?', '#' => true,
+        else => false,
+    };
+}
+
+fn isValidMacAddress(mac_address: []const u8) bool {
+    // A valid mac address has 6 two-character components with 5 colons, e.g. 12:34:56:ab:cd:ef.
+    if (mac_address.len != 17) {
+        return false;
+    }
+
+    for (0..mac_address.len) |i| {
+        const c = mac_address[i];
+
+        if ((i + 1) % 3 == 0) {
+            if (c != ':') {
+                return false;
+            }
+        } else {
+            if (!std.mem.containsAtLeastScalar(u8, "0123456789ABCDEFabcdef", 1, c)) {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+/// Parses the provided url to a `std.Uri` struct. This is very specific to getting hostname and
+/// path information for Ghostty's PWD reporting functionality. Takes into account that on macOS
+/// the url passed to this function might have a mac address as its hostname and parses it
+/// correctly.
+pub fn parseUrl(url: []const u8) !std.Uri {
+    return std.Uri.parse(url) catch |e| {
+        // It's possible this is a mac address on macOS where the last 2 characters in the
+        // address are non-digits, e.g. 'ff', and thus an invalid port.
+        //
+        // Example: file://12:34:56:78:90:12/path/to/file
+        if (e != error.InvalidPort) return e;
+
+        const scheme, const url_without_scheme = url: {
+            if (std.mem.startsWith(u8, url, "file://")) break :url .{ "file", url[7..] };
+            if (std.mem.startsWith(u8, url, "kitty-shell-cwd://")) break :url .{
+                "kitty-shell-cwd",
+                url[18..],
+            };
+
+            return error.UnsupportedScheme;
+        };
+
+        // The first '/' after the scheme marks the end of the hostname. If the first '/'
+        // following the end of the scheme is not at the right position this is not a
+        // valid mac address.
+        if (std.mem.indexOfScalarPos(u8, url_without_scheme, 0, '/') != 17 and
+            url_without_scheme.len != 17)
```

Comment:
Could `url_without_scheme.len != 17` be evaluated first? Seems like a cheap condition to check to see if we fall into the MAC address case.

The `17`s also seem like a magic number we could consider naming as a constant.

Reply:
Oh, nice catch! Fixed the if-statement order in 550321375fe80573a9acbe9712440338efed478d and renamed the magic constant in 5fa0fb0a1c7c9a314ce51214d833ca9c078a8970 🙌